src/client/Microsoft.Identity.Client/OAuth2/TokenClient.cs

```diff
@@ -161,6 +161,54 @@ private void AddBodyParamsAndHeaders(IDictionary<string, string> additionalBodyP
             {
                 _oAuth2Client.AddHeader(PKeyAuthConstants.DeviceAuthHeaderName, PKeyAuthConstants.DeviceAuthHeaderValue);
             }
+
+            AddCCSRoutingInformation(additionalBodyParameters);
+
+            AddExtraHttpHeaders();
+        }
+
+        private void AddCCSRoutingInformation(IDictionary<string, string> additionalBodyParameters)
+        {
+            if (_requestParams.Account != null && _requestParams.Account.HomeAccountId != null)
+            {
+                if (!String.IsNullOrEmpty(_requestParams.Account.HomeAccountId.Identifier))
+                {
+                    var userObjectId = _requestParams.Account.HomeAccountId.ObjectId;
+                    var userTenantID = _requestParams.Account.HomeAccountId.TenantId;
+                    string OidCCSHeader = $@":""oid:<{userObjectId}>@<{userTenantID}>""";
+
+                    _oAuth2Client.AddHeader(Constants.OidCCSHeader, OidCCSHeader);
+                }
+                else if (!String.IsNullOrEmpty(_requestParams.Account.Username))
+                {
+                    AddCCSUpnHeader(_requestParams.Account.Username);
+                }
+            }
+            else if (additionalBodyParameters.ContainsKey(OAuth2Parameter.Username))
+            {
+                AddCCSUpnHeader(additionalBodyParameters[OAuth2Parameter.Username]);
+            }
+            else if (!String.IsNullOrEmpty(_requestParams.LoginHint))
+            {
+                AddCCSUpnHeader(_requestParams.LoginHint);
+            }
+        }
+
+        private void AddExtraHttpHeaders()
+        {
+            if (_requestParams.ExtraHttpHeaders != null)
+            {
+                foreach (KeyValuePair<string, string> pair in _requestParams.ExtraHttpHeaders)
+                {
+                    _oAuth2Client.AddHeader(pair.Key, pair.Value);
```

Comment:
What happens if the user adds a header which already exists, for example x-client-sku ? IMO the user value should overwrite the lib value. We should have a test for this.

Reply:
The code in the OAuth client will overwrite the previous value if the key for the header already exists.
see https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/blob/e6c3d8cb66232dd69a7f0229b24ffe171a8edf8c/src/client/Microsoft.Identity.Client/OAuth2/OAuth2Client.cs#L62